File: examples/readers/python/reader.py
"""Reference Python reader for emit's state.snap files.

Decodes a MutableStore's records without any emit-sdk dependency. Usage
against the ERC20 example:

    python3 reader.py /path/to/erc20/data/state.snap

The schema (number of mutables, number of immutables, per-slot record size,
field layout) is comptime-known in the indexer. Readers consume that as
external knowledge — there is no per-slot descriptor on disk. See
docs/entity-format.md for the byte layout.
"""

import struct
import sys
from dataclasses import dataclass

# ERC20 schema. Slot 0: Account = id [20]u8 + balance u256.
# Slot 1: Allowance = id [40]u8 + value u256.
NUM_MUTABLE = 2
NUM_IMMUTABLE = 2  # Transfer, Approval -- counts only; records live in events.dat files.
ACCOUNT_RECORD_SIZE = 20 + 32

@dataclass
class Account:
    id: bytes      # 20 bytes (BE since it's the primary key)
    balance: int   # u256 LE

def read_state_snap(path: str):
    with open(path, "rb") as f:
        buf = f.read()

    magic, version, cursor = struct.unpack_from("<8s I Q", buf, 0)
    assert magic == b"EMITSTAT", f"bad magic: {magic!r}"
    assert version == 1, f"unsupported version: {version}"

    pos = 8 + 4 + 8
    mutable_bytes = struct.unpack_from(f"<{NUM_MUTABLE}Q", buf, pos)
    pos += NUM_MUTABLE * 8
    immutable_counts = struct.unpack_from(f"<{NUM_IMMUTABLE}Q", buf, pos)
    pos += NUM_IMMUTABLE * 8

    # Body: slabs concatenated in slot order. Slot 0 = Account slab.
    account_slab = buf[pos : pos + mutable_bytes[0]]
    assert len(account_slab) % ACCOUNT_RECORD_SIZE == 0
    accounts = []
    for i in range(0, len(account_slab), ACCOUNT_RECORD_SIZE):
        record = account_slab[i : i + ACCOUNT_RECORD_SIZE]
        # First field is BE per the format spec (sorted-by-bytes order).
        id_bytes = record[0:20]
        # u256 is little-endian for the data fields.
        balance = int.from_bytes(record[20:52], "little")
        accounts.append(Account(id=id_bytes, balance=balance))

    return cursor, accounts, immutable_counts

if __name__ == "__main__":
    cursor, accounts, immutable_counts = read_state_snap(sys.argv[1])
    print(f"cursor: {cursor}")
    print(f"accounts: {len(accounts)}")
    print(f"immutable counts (Transfer, Approval): {immutable_counts}")
    if accounts:
        a = accounts[0]
        print(f"first account: id=0x{a.id.hex()} balance={a.balance}")

File: examples/readers/zig/reader.zig
/// Reference Zig reader for emit's state.snap files. No dependency on emit-sdk.
///
/// Decodes a MutableStore's records straight from
/// the binary format documented in docs/entity-format.md.
///
/// Usage against the ERC20 example:
///
///     zig build-exe reader.zig -O ReleaseFast
///     ./reader /path/to/erc20/data/state.snap
///
/// The schema (mutable count, immutable count, per-slot record size, field
/// layout) is comptime-known in the indexer. Readers supply it as external
/// knowledge. No per-slot descriptor exists on disk.
const std = @import("std");

const MAGIC: *const [8]u8 = "EMITSTAT";
const VERSION: u32 = 1;

// ERC20 schema. Adjust these constants for other indexers.
const NUM_MUTABLE = 2; // Account, Allowance
const NUM_IMMUTABLE = 2; // Transfer, Approval

const Account = struct {
    id: [20]u8, // big-endian (primary key)
    balance: u256, // little-endian

    const RECORD_SIZE: usize = 20 + 32;
};

pub fn main() !void {
    var gpa: std.heap.GeneralPurposeAllocator(.{}) = .{};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const args = try std.process.argsAlloc(allocator);
    defer std.process.argsFree(allocator, args);
    if (args.len != 2) {
        std.debug.print("usage: {s} <state.snap>\n", .{args[0]});
        return error.MissingArg;
    }

    const file = try std.fs.cwd().openFile(args[1], .{});
    defer file.close();
    const stat = try file.stat();
    const buf = try allocator.alloc(u8, stat.size);
    defer allocator.free(buf);
    if ((try file.readAll(buf)) != stat.size) return error.Truncated;

    if (!std.mem.eql(u8, buf[0..8], MAGIC)) return error.InvalidMagic;
    const version = std.mem.readInt(u32, buf[8..12], .little);
    if (version != VERSION) return error.VersionMismatch;
    const cursor = std.mem.readInt(u64, buf[12..20], .little);

    var pos: usize = 20;
    var mutable_bytes: [NUM_MUTABLE]u64 = undefined;
    for (&mutable_bytes) |*b| {
        b.* = std.mem.readInt(u64, buf[pos..][0..8], .little);
        pos += 8;
    }
    var immutable_counts: [NUM_IMMUTABLE]u64 = undefined;
    for (&immutable_counts) |*c| {
        c.* = std.mem.readInt(u64, buf[pos..][0..8], .little);
        pos += 8;
    }

    // Slot 0 = Account slab. Slot 1 = Allowance, not decoded here (same
    // struct/parse pattern with a different layout).
    const account_slab = buf[pos .. pos + mutable_bytes[0]];
    std.debug.assert(account_slab.len % Account.RECORD_SIZE == 0);
    const account_count = account_slab.len / Account.RECORD_SIZE;

    std.debug.print("cursor:            {d}\n", .{cursor});
    std.debug.print("accounts:          {d}\n", .{account_count});
    std.debug.print("transfer count:    {d}\n", .{immutable_counts[0]});
    std.debug.print("approval count:    {d}\n", .{immutable_counts[1]});

    if (account_count > 0) {
        const first = parseAccount(account_slab[0..Account.RECORD_SIZE]);
        std.debug.print("first account:     id=0x", .{});
        for (first.id) |b| std.debug.print("{x:0>2}", .{b});
        std.debug.print(" balance={d}\n", .{first.balance});
    }
}

fn parseAccount(record: *const [Account.RECORD_SIZE]u8) Account {
    // First field is big-endian (sorted-by-bytes order). Data fields are
    // little-endian.
    return .{
        .id = record[0..20].*,
        .balance = std.mem.readInt(u256, record[20..52], .little),
    };
}
